File: Tugas Besar 1/Animal/Goat.cpp
/**
 * @file Goat.cpp
 * @author 13515125/Muhammad Rafid Amrullah, 13517092/Michael Ray, 13517113/Juro Sutantra, 13517137/Vincent Budianto
 * @brief file berisi implementasi class Goat
 * @version 0.1
 * @date 2019-03-20
 */

/* NIM/Nama  : 13515125/Muhammad Rafid Amrullah, 13517092/Michael Ray, 13517113/Juro Sutantra, 13517137/Vincent Budianto
 * Nama file : Goat.cpp
 * Topik     : Tugas Besar 1 IF2210 - Pemograman Berorientasi Objek
 * Tanggal   : 20 Maret 2019
 * Deskripsi : Implementasi Goat.cpp */

#include "Goat.hpp"
#include <string.h>
#include <random>
#include <iostream>
using namespace std;

int Goat::n_goat = 0;

/**
 * @brief Construct a new Goat object
 * 
 */
Goat::Goat(int id, int x, int y)
{
    this->id = id;
    this->voice = new char[15];
    strcpy(this->voice, "Mbee!");
    this->hungry = false;
    this->x = x;
    this->y = y;
    this->tick = 0;
    this->starving = false;
    this->interactivity = true;
    n_goat++;
}

/**
 * @brief Destroy the Goat object
 * 
 */
Goat::~Goat()
{

}

/**
 * @brief Get the Hungry object
 * 
 * @return boolean 
 */
bool Goat::getHungry() const
{
    return this->hungry;
}

/**
 * @brief Get the X object
 * 
 * @return int 
 */
int Goat::getX()
{
    return this->x;
}

/**
 * @brief Get the Y object
 * 
 * @return int 
 */
int Goat::getY()
{
    return this->y;
}

/**
 * @brief Function to get interactivity
 * 
 */
bool Goat::getInteractivity()
{
    return interactivity;
}

/**
 * @brief Function to set interactivity
 * 
 */
void Goat::setInteractivity(bool stat)
{
    interactivity = stat;
}

/**
 * @brief Method for the animal to eat
 * 
 */
void Goat::eat()
{
    this->hungry = false;
    this->interactivity = true;
    this->tick = 0;
}

/**
 * @brief Method for the animal to move
 * 
 */
void Goat::move(char** map, int row, int col)
{
    int turn;
    char next = '.';

    random_device dev;
    mt19937 rng(dev());
    uniform_int_distribution<mt19937::result_type> dist6(1,4);
    turn = dist6(rng);

    if (turn == 1)
    {
        if(this->y + 1 < row)
            next = map[this->y+1][this->x];
        if (next == 'x' or next == '@')
        {
            this->y++;
        }
    }
    else if (turn == 2)
    {
        if(this->x + 1 < col)
            next = map[this->y][this->x+1];
        if (next == 'x' or next == '@')
        {
            this->x++;
        }
    }
    else if (turn == 3)
    {
        if(this->y - 1 >= 0)
            next = map[this->y-1][this->x];
        if (next == 'x' or next == '@')
        {
            this->y--;
        }
    }
    else if (turn == 4)
    {
        if(this->x - 1 >= 0)
            next = map[this->y][this->x-1];
        if (next == 'x' or next == '@')
        {
            this->x--;
        }
    }

    if(tick == 5){
        hungry = true;
    }else if(tick >= 10 and hungry){
        starving = true;
    }
    tick++;
}

/**
 * @brief Method to get starvation status
 * 
 */
bool Goat::getStarvation()
{
    return starving;
}

/**
 * @brief Method for the animal to voice
 * 
 */
void Goat::sound()
{
    cout << this->voice << endl;
}

/**
 * @brief Method to render the animal to map
 * 
 * @return char 
 */
char Goat::render()
{
    if (this->hungry)
    {
		return 'g';
	}
	else
	{
		return 'G';
	}
}

File: Java/model/animal/Goat.java
/**
 * Goat.java
 * <p>
 * File berisi class Goat
 * </p>
 * @author 13515125/Muhammad Rafid Amrullah, 13517113/Juro Sutantra, 13517137/Vincent Budianto
 * @version 2.0
 * @since 2019-04-21
 */

package animal;

import java.io.*;
import java.lang.*;
import java.util.*;

/**
 * <p>
 * Class for rendering object to Goat
 * </p>
 */
public class Goat implements KProducing
{
	private static int n_goat;
	private int id;
	private String voice;
	private boolean hunger, starvation, interactivity, produceMeat;
	private int x, y, tick;

	/**
	 * Goat constructor
	 * @param y
	 * @param x
	 */
	public Goat(int y, int x){
		this.id = ++n_goat;
		this.voice = "Mbeee";
		this.hunger = false;
		this.y = y;
		this.x = x;
		this.tick = 0;
		this.starvation = false;
		this.interactivity = true;
		this.produceMeat = false;
	}

	/**
	 * Get the object id
	 * @return int
	 */
	public int getId(){
		return id;
	}

	/**
	 * Get the object hunger status
	 * @return boolean
	 */
	public boolean getHunger(){
		return hunger;
	}

	/**
	 * Get the object interactivity
	 * @return boolean
	 */
	public boolean getInteractivity(){
		return interactivity;
	}

	/**
	 * Get the object starve status
	 * @return boolean
	 */
	public boolean getStarvation(){
		return starvation;
	}

	/**
	 * Get the object kill status
	 * @return boolean
	 */
	public boolean isMeatProducing(){
		return produceMeat;
	}

	/**
	 * Get the object x position
	 * @return int
	 */
	public int getX(){
		return this.x;
	}

	/**
	 * Get the object y position
	 * @return int
	 */
	public int getY(){
		return this.y;
	}

	/**
	 * Eat function
	 */
	public void eat(){
		hunger = false;
		interactivity = true;
		tick = 0;
	}

	/**
	 * Move function
	 * @param map
	 */
	public void move(Character[][] map){
		int row = map.length;
		int col = map[0].length;
	    char next = '.';

	    Random random = new Random();
	    int turn = random.nextInt(5);

	    if (turn == 1)
	    {
	        if(this.y + 1 < row)
	            next = map[this.y+1][this.x];
	        if (next == 'x' || next == '@')
	        {
	            this.y++;
	        }
	    }
	    else if (turn == 2)
	    {
	        if(this.x + 1 < col)
	            next = map[this.y][this.x+1];
	        if (next == 'x' || next == '@')
	        {
	            this.x++;
	        }
	    }
	    else if (turn == 3)
	    {
	        if(this.y - 1 >= 0)
	            next = map[this.y-1][this.x];
	        if (next == 'x' || next == '@')
	        {
	            this.y--;
	        }
	    }
	    else if (turn == 4)
	    {
	        if(this.x - 1 >= 0)
	            next = map[this.y][this.x-1];
	        if (next == 'x' || next == '@')
	        {
	            this.x--;
	        }
	    }

	    boolean validTile = map[this.y][this.x] != '-' && map[this.y][this.x] != '#';
	    validTile = validTile && map[this.y][this.x] != 'x' && map[this.y][this.x] != '@';

	    if(validTile)
	    	move(map);

	    if(tick == 5)
	        hunger = true;
	    else if(tick >= 25 && hunger)
	        starvation = true;

	    tick++;
	}

	/**
	 * Get the object sound
	 * @return String
	 */
	public String sound(){
		return new String("Goat " + getId() + ": " + voice);
	}

	/**
	 * Get the object symbol
	 * @return char
	 */
	public char render(){
		return (hunger)? 'g' : 'G';
	}

	/**
	 * Interact function
	 */
	public void interact(){
		interactivity = false;
	}

	/**
	 * Kill function
	 * @return String
	 */
	public String kill(){
		return ("Goat " + getId() + " is killed");
	}
}
